### DISCUSSION TITLE
contributing/development/compiling/compiling_for_macos

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#197] [User-contributed notes]
# contributing/development/compiling/compiling_for_macos

Requirements: For compiling under macOS, the following is required: Python 3.6+., SCons 3.1.2+ build system., Xcode(or the more lightweight Command Line Tools for Xcode)., Vulkan SDK for MoltenVK (...

https://docs.godotengine.org/en/latest/contributing/development/compiling/compiling_for_macos.html

**Comment by jessdecker** [comment]
When building on macOS and generating the export templates, I struggled a lot until I found a flag called "generate_bundle".  Ultimately, this worked: "scons platform=macos arch=x86_64 target=template_release module_mono_enabled=yes precision=double generate_bundle=true"

**Reply by Calinou** [reply]
Pull request opened:

- https://github.com/godotengine/godot-docs/pull/10096

**Comment by OS-of-S** [comment]
I tried it for Ubuntu 20.04 (and couldn't solve the last problem that occurs already during compilation in the last stage of linking, but never mind...) It required to install MoltenVK SDK, wich aren't mentioned on this page. And it's not only my problem: https://www.reddit.com/r/godot/comments/1af8nk1/compiling_on_macos/

After experiments with Ubuntu 20.04 I checed dockers for Fedora 36, and there I was needed to download MoltenVK separately and use scons command as
scons platform=macos vulkan_sdk_path=$HOME/MoltenVK osxcross_sdk=darwin24

I just not completly sure, why isn't it mentioned, and which way of MoltenVK installing better and official. I also didn't checked if compiled Godot works, but compilation itself finishes. I think it should be mentioned in the page in some form.

**Reply by OS-of-S** [reply]
Also it would be nice to mention, that osxcross_sdk you can take from filenames inside $OSXCROSS_ROOT/target/bin, and maybe osxcross_sdk option needs to be positioned as necessary, since in almost any case the user's version will not match the one that scons expects. In the page aren't even mentioned, which XCode version recommended.

**Reply by OS-of-S** [reply]
At least, I want to say that I wrote down all steps I made on Linux in this script: https://github.com/OS-of-S/Godot_LinuxLibsInstaller

There is a "MacOS-part" after "Linux-part". Maybe it whould be useful to someone.

**Comment by okachobi** [comment]
A word of advice that may save you time - if you have gnu binutils installed from brew in your path, it may cause some issues with the build.  I encountered a file not found for "__algorithm/copy_backwards.h" (which was present in the MacOS SDK- but not being found for some reason) and an issue with linking the static .a libs due to what I think was the build picking up ar from the binutils.  I removed binutils from my PATH and the build worked as expected.

**Reply by sam-omalley** [reply]
Thanks @okachobi. I also had trouble with building on osx (Apple Silicon) that was caused by homebrew's binutils.

If anyone sees an error like `archive member is not a mach-o file` then for me it was caused by using `/opt/homebrew/opt/binutils/bin/ar`, rather than `/usr/bin/ar`. This was fixed by `brew remove binutils`

**Comment by MaxKablaam** [comment]
I noticed that there is a handy script `engine/godot/misc/scripts/install_vulkan_sdk_macos.sh` you can run to install Vulkan SDK. Might be worth documenting that instead of, or along with, linking to the direct download.

**Reply by Calinou** [reply]
Pull request opened:

- https://github.com/godotengine/godot-docs/pull/10465

**Comment by TCRodriguez** [comment]
This is probably an incredibly minor (and obvious) detail, but for the line:
```
cp bin/godot.macos.editor.universal Godot.app/Contents/MacOS/Godot
```
It may be more clear to specify that in the first part of the command (`cp bin/godot.macos.editor.universal`) the `universal` part could also either be `x86_64`, `arm64` depending on what was passed in during the initial `scons` compile call. 

It's my first time building something from source so I'm stepping through things inch by inch to learn and understand :)

**Reply by TCRodriguez** [reply]
Forgot to mention that this is under the `Compiling` section ><.

**Comment by hardaker** [comment]
Note that for Vulkan you need an older version, per https://www.reddit.com/r/godot/comments/1af8nk1/compiling_on_macos/

**Reply by Calinou** [reply]
It shouldn't be necessary to use an older MoltenVK to build the `master` branch of Godot, or even `4.4` for that matter. Please open an issue on https://github.com/godotengine/godot if you can't compile with MoltenVK after running `misc/scripts/install_vulkan_sdk_macos.sh`.

**Reply by hardaker** [reply]
I should have mentioned this is for cross-compiling from linux, where `scons` fails after running `misc/scripts/install_vulkan_sdk_macos.sh` because it's looking for `libMoltenVK.a` which doesn't exist in newer versions.
